### PR TITLE
Add DHCPv6 Relay Agent

### PR DESCRIPTION
**Pull request by kellyyeh**
DHCPv6 is a network protocol used to assign IPv6 addresses and provide configuration for devices to communicate on a network. DHCPv6 relay agent helps forwards DHCP packets between clients and servers that do not reside on a shared physical subnet, so that all subnets can share DHCPv6 server, and DHCPv6 server is not required on every LAN. This DHCPv6 relay agent supports option 79 which incorporates the client link-layer address in Relay-Forward messages that is sent to the server, which is needed by the server to allow IPv6 reservation to clients based on their physical MAC addresses.

This PR is part of: https://github.com/Azure/SONiC/pull/787
Signed-off-by: Kelly Yeh kellyyeh@microsoft.com

#### Why I did it
To support option 79 in DHCPv6.

#### How I did it
Implemented DHCPv6 operations along with option 79.
Read from config_db to process dhcpv6 server addresses and options.

#### How to verify it
Build the binary with Makefile and run it on switch with IPv6 configured and DHCPv6 server support

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [X] 202012
- [ ] 202106

### CODE REVIEW

File: src/dhcp6relay/src/configInterface.cpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+ */
+void processRelayNotification(std::deque<swss::KeyOpFieldsValuesTuple> &entries, arg_config *context)
+{
+    std::vector<std::string> servers;
+
+    for (auto &entry: entries) {
+        std::string vlan = kfvKey(entry);
+        std::vector<swss::FieldValueTuple> fieldValues = kfvFieldsValues(entry);
+
+        for (auto &fieldValue: fieldValues) {
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/configInterface.cpp:104]: we need to check the operation here if it is ADD/DEL

File: src/dhcp6relay/src/main.cpp [hunk 2]
```diff
@@ -0,0 +1,18 @@
+#include <stdlib.h>
+#include "configInterface.h"
+
+int main(int argc, char *argv[]) {
+    try {
+        auto parser = ArgumentParser();
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/main.cpp:6]: do we need arg parser now?

File: src/dhcp6relay/src/relay.cpp [hunk 3]
```diff
@@ -0,0 +1,665 @@
...
+
+/**
+ * @code        parse_args(int argc, char *argv[]);
+ *
+ * @brief       parse through argument and define the interface to bind to and to turn off option 79. 
+ *              read from config database and set the ipv6 helper addresses.
+ *
+ * @return      none
+ */
+void ArgumentParser::parse_args(int argc, char *argv[]) {
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/relay.cpp:33]: is this required?

File: src/dhcp6relay/src/relay.cpp [hunk 4]
```diff
@@ -0,0 +1,665 @@
...
+
+    for(auto server: context->servers) {
+        sockaddr_in6 tmp;
+        if(inet_pton(AF_INET6, server.c_str(), &tmp.sin6_addr) != 1)
+        {
+            syslog(LOG_WARNING, "inet_pton: Failed to convert IPv6 address\n");
+        }
+        tmp.sin6_family = AF_INET6;
+        tmp.sin6_flowinfo = 0;
+        tmp.sin6_port = htons(547);
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/relay.cpp:219]: named constant for port 547

File: src/dhcp6relay/src/relay.cpp [hunk 5]
```diff
@@ -0,0 +1,665 @@
...
+
+    sockaddr_in6 *non_link_local = NULL;    
+    sockaddr_in6 *link_local = NULL;
+    
+    struct ifaddrs *ifa, *ifa_tmp;
+    char addr[INET6_ADDRSTRLEN];
+    addr[0] = 0;
+
+    if (getifaddrs(&ifa) == -1) {
+		    syslog(LOG_WARNING, "getifaddrs: Unable to get network interfaces\n");
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/relay.cpp:232]: nit: use spaces only for indentation

File: src/dhcp6relay/src/relay.cpp [hunk 6]
```diff
@@ -0,0 +1,665 @@
...
+            }
+            if((strcmp(ifa_tmp->ifa_name, interface_config->interface.c_str()) == 0) && is_addr_link_local(in6->sin6_addr)) {    
+                link_local = in6;
+            }   
+	    }
+	    ifa_tmp = ifa_tmp->ifa_next;
+	}
+	freeifaddrs(ifa);
+
+    if (non_link_local != NULL) {
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/relay.cpp:252]: what about multiple link local and/or global addresses?

File: src/dhcp6relay/src/relay.cpp [hunk 7]
```diff
@@ -0,0 +1,665 @@
...
+            default:
+                break;
+        }
+    }
+
+    strcpy(ifname, configs->interface.c_str());
+    memcpy(&target_addr.sin6_addr, &dhcp_relay_header->peer_address, sizeof(struct in6_addr));
+    target_addr.sin6_family = AF_INET6;
+    target_addr.sin6_flowinfo = 0;
+    target_addr.sin6_port = htons(546);
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/relay.cpp:403]: same for port 546

File: src/dhcp6relay/src/relay.cpp [hunk 8]
```diff
@@ -0,0 +1,665 @@
...
+ *
+ * @param fd            filter socket
+ * @param event         libevent triggered event  
+ * @param arg           callback argument provided by user
+ *
+ * @return              none
+ */
+void callback(evutil_socket_t fd, short event, void *arg) {
+    struct relay_config *config = (struct relay_config *)arg;
+    uint8_t message_buffer[4096];
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/relay.cpp:423]: we can make this buffer static.

File: src/dhcp6relay/src/relay.cpp [hunk 9]
```diff
@@ -0,0 +1,665 @@
...
+        while (ext_header->ip6e_nxt != IPPROTO_UDP);
+    }  
+
+    auto udp_header = parse_udp(current_position, &tmp);
+    current_position = tmp;
+
+    auto dhcp_header = parse_dhcpv6_hdr(current_position);
+
+    if ((ntohs(udp_header->dest)) == 547 ) {
+        if(dhcp_header->msg_type == SOLICIT || REQUEST || CONFIRM || RENEW || REBIND || RELEASE || DECLINE) {
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/relay.cpp:454]: what is meant by this condition? this is true all the time

File: src/dhcp6relay/src/relay.cpp [hunk 10]
```diff
@@ -0,0 +1,665 @@
...
+ * @code                listen_server(void *arg);
+ * 
+ * @brief               listen to the DHCPv6 server
+ *
+ * @param arg           callback argument provided by user
+ *
+ * @return              none
+ */
+void listen_server(relay_config *config) {
+    while (serverListener) {
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/relay.cpp:470]: why the server listener is not part of the even loop?

**Review by tahmed-dev — COMMENTED** [review]
Thanks!

File: src/dhcp6relay/src/main.cpp (hunk 2 above)
Comment (kellyyeh) [reply src/dhcp6relay/src/main.cpp]: Deleted

File: src/dhcp6relay/src/relay.cpp (hunk 3 above)
Comment (kellyyeh) [reply src/dhcp6relay/src/relay.cpp]: Deleted along with arg parser.

File: src/dhcp6relay/src/relay.cpp (hunk 6 above)
Comment (kellyyeh) [reply src/dhcp6relay/src/relay.cpp:207]: Any global address will do. Link local address is only used if not a single global address is present

File: src/dhcp6relay/src/relay.cpp (hunk 9 above)
Comment (kellyyeh) [reply src/dhcp6relay/src/relay.cpp]: Thanks for catching. Fixed.

File: src/dhcp6relay/src/configInterface.cpp (hunk 1 above)
Comment (kellyyeh) [reply src/dhcp6relay/src/configInterface.cpp:104]: The dhcpv6 server address are type string value. e.g.: "fc02:2000::1,fc02:2000::2"
HDEL would delete the entire string of addresses. HSET updates the list of server addresses.
When config_db updates the list of server addresses, all addresses are properly fetched and processed regardless new addresses are added/deleted. Therefore, I don't think there is a need to check for add/del

File: src/dhcp6relay/src/relay.cpp [hunk 11]
```diff
@@ -450,10 +405,8 @@ void callback(evutil_socket_t fd, short event, void *arg) {
 
     auto dhcp_header = parse_dhcpv6_hdr(current_position);
 
-    if ((ntohs(udp_header->dest)) == 547 ) {
-        if(dhcp_header->msg_type == SOLICIT || REQUEST || CONFIRM || RENEW || REBIND || RELEASE || DECLINE) {
-            relay_client(config->local_sock, current_position, ntohs(udp_header->len) - sizeof(udphdr), ip_header, ether_header, config);
-        }    
+    if (((ntohs(udp_header->dest)) == RELAY_PORT) && (dhcp_header->msg_type == SOLICIT || REQUEST || CONFIRM || RENEW || REBIND || RELEASE || DECLINE)) {
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/relay.cpp:408]: This condition will always be true:
`(dhcp_header->msg_type == SOLICIT || REQUEST || CONFIRM || RENEW || REBIND || RELEASE || DECLINE)`

File: src/dhcp6relay/src/configInterface.cpp (hunk 1 above)
Comment (tahmed-dev) [reply src/dhcp6relay/src/configInterface.cpp:104]: You can get the operation using `std::string operation = kfvOp(entry);`. This way we can tell the type of the record updates. I think they are `SET/DEL` operation.

File: src/dhcp6relay/src/relay.cpp (hunk 10 above)
Comment (kellyyeh) [reply src/dhcp6relay/src/relay.cpp]: Added server listener to libevent

File: src/dhcp6relay/src/configInterface.cpp [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+ * @brief               main thread for handling SWSS notification
+ *
+ * @param context       argument config that contains strings of server and option
+ *
+ * @return              none
+ */
+void handleSwssNotification(arg_config *context)
+{
+    std::shared_ptr<swss::DBConnector> configDbPtr = std::make_shared<swss::DBConnector> ("CONFIG_DB", 0);
+    swss::SubscriberStateTable ipHelpersTable(configDbPtr.get(), "DHCP");
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/configInterface.cpp:49]: We need to add `DHCP` as a table name in `schema.h` file.

File: src/dhcp6relay/src/configInterface.cpp [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+                    getline(ss, substr, ',');
+                    context->servers.push_back(substr);
+                }
+            }
+            if(f == "options") {
+                std::stringstream ss(v);
+                while (ss.good()) {
+                    std::string substr;
+                    getline(ss, substr, ',');
+                    if(substr == "79")
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/configInterface.cpp:120]: is the option 79 name up-to-date?

File: src/dhcp6relay/src/configInterface.h [hunk 14]
```diff
@@ -0,0 +1,16 @@
+#include <boost/thread.hpp>
+#include "subscriberstatetable.h"
+#include "select.h"
+#include "relay.h"
+
+void initialize_swss(arg_config *context);
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/configInterface.h:6]: please add doxygen comments to functions in this file.

File: src/dhcp6relay/src/relay.cpp [hunk 15]
```diff
@@ -0,0 +1,588 @@
...
+
+    sockaddr_in6 *non_link_local = NULL;    
+    sockaddr_in6 *link_local = NULL;
+    
+    struct ifaddrs *ifa, *ifa_tmp;
+    char addr[INET6_ADDRSTRLEN];
+    addr[0] = 0;
+
+    if (getifaddrs(&ifa) == -1) {
+		syslog(LOG_WARNING, "getifaddrs: Unable to get network interfaces\n");
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/relay.cpp:184]: nit: indentations are off. Please use space for indentation

File: src/dhcp6relay/src/relay.cpp [hunk 16]
```diff
@@ -0,0 +1,588 @@
...
+        exit(1);
+	}
+
+    ifa_tmp = ifa;
+	while (ifa_tmp) {
+		if ((ifa_tmp->ifa_addr) && (ifa_tmp->ifa_addr->sa_family == AF_INET6)) {
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/relay.cpp:190]: indentation

File: src/dhcp6relay/src/main.cpp [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+
+int main(int argc, char *argv[]) {
+    try {
+        arg_config context;
+        initialize_swss(&context);
+        loop_relay(&context);
+    }
+    catch (std::exception &e)
+    {
+        printf("Usage: -i <interface> -o <option>\n");
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/main.cpp:12]: nit: we are not using cli args anymore.

File: src/dhcp6relay/src/relay.h [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+    ArgumentException(std::string message) : msg(message) {}
+    ~ArgumentException() throw() {}
+    const char* what() const throw() { return msg.c_str(); }
+};
+
+
+/* DHCPv6 filter */
+/* sudo tcpdump -dd "ip6 dst ff02::1:2 && udp dst port 547" */
+
+static struct sock_filter ether_relay_filter[] = {
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/relay.h:138]: please move this static variable to .c file

File: src/dhcp6relay/src/relay.cpp [hunk 19]
```diff
@@ -0,0 +1,588 @@
...
+    if (((ntohs(udp_header->dest)) == RELAY_PORT) && (dhcp_header->msg_type == SOLICIT || REQUEST || CONFIRM || RENEW || REBIND || RELEASE || DECLINE)) {
+        relay_client(config->local_sock, current_position, ntohs(udp_header->len) - sizeof(udphdr), ip_header, ether_header, config);
+    }
+}
+
+void server_callback(evutil_socket_t fd, short event, void *arg) {
+    struct relay_config *config = (struct relay_config *)arg;
+    sockaddr_in6 from;
+    int32_t data = 0;
+    uint8_t message_buffer[4096];
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/relay.cpp:432]: can we make `message_buffer` static?

File: src/dhcp6relay/src/main.cpp [hunk 20]
```diff
@@ -0,0 +1,16 @@
+#include <stdlib.h>
+#include "configInterface.h"
+
+int main(int argc, char *argv[]) {
+    try {
+        arg_config context;
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/main.cpp:6]: is `context` a per vlan struct? How are we handling multi-vlan case?

Comment (kellyyeh) [reply src/dhcp6relay/src/main.cpp]: Updated code to support multi-vlan

File: src/dhcp6relay/src/main.cpp [hunk 21]
```diff
@@ -0,0 +1,17 @@
+#include <stdlib.h>
+#include "configInterface.h"
+
+int main(int argc, char *argv[]) {
+    try {
+        std::vector<arg_config> vlans;
+        swss::DBConnector state_db(6, "localhost", 6379, 0);
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/main.cpp:7]: you can use `swss::DBConnector("STATE_DB", 0)` constructor.

File: src/dhcp6relay/src/relay.cpp [hunk 22]
```diff
@@ -0,0 +1,672 @@
...
+}
+
+std::string toString(uint16_t count) {
+    std::stringstream ss;
+    ss << count;
+    std::string countValue = ss.str();
+    return countValue;
+}
+
+bool is_addr_gua(in6_addr addr) {
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/relay.cpp:83]: code comments for those methods

File: src/dhcp6relay/src/relay.cpp (hunk 16 above)
Comment (tahmed-dev) [reply src/dhcp6relay/src/relay.cpp:250]: please remove tabs and use 4 spaces indentation.

File: src/dhcp6relay/src/relay.cpp [hunk 23]
```diff
@@ -0,0 +1,672 @@
...
+ *
+ * @param sock          L3 socket for sending data to servers
+ * @param msg           pointer to dhcpv6 message header position
+ * @param len           size of data received
+ * @param config        relay interface config
+ *
+ * @return              none
+ */
+ void relay_relay_reply(int sock, const uint8_t *msg, int32_t len, relay_config *configs) {
+    uint8_t buffer[4096];
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/relay.cpp:417]: can you please avoid having large local vars?

File: src/dhcp6relay/src/main.cpp [hunk 24]
```diff
@@ -4,12 +4,13 @@
...
-        initialize_swss(&vlans);
-        loop_relay(&vlans);
+        swss::DBConnector state_db(6, "localhost", 6379, 0);
+        initialize_swss(&vlans, &state_db);
+        loop_relay(&vlans, &state_db);
     }
     catch (std::exception &e)
     {
-        printf("Usage: -i <interface> -o <option>\n");
+        printf("An exception occurred.\n");
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/main.cpp:13]: please use `syslog`

File: src/dhcp6relay/src/relay.cpp [hunk 25]
```diff
@@ -20,6 +20,66 @@ struct event_base *base;
...
+	{ 0x15, 0, 1, 0x00000223 },
+	{ 0x6, 0, 0, 0x00040000 },
+	{ 0x6, 0, 0, 0x00000000 },
+};
+const struct sock_fprog ether_relay_fprog = {
+	lengthof(ether_relay_filter),
+	ether_relay_filter
+};
+
+/* DHCPv6 Counter */
```
Comment (tahmed-dev) [inline src/dhcp6relay/src/relay.cpp:50]: can we split counters to be per vlan?

Comment (kellyyeh) [reply src/dhcp6relay/src/relay.cpp:49]: Yes. Updated: Each vlan will have its own counter table (ex: DHCPv6_Counter|Vlan1000, DHCPv6_Counter|Vlan2000)

**Review by tahmed-dev — DISMISSED** [review]

**Review by tahmed-dev — DISMISSED** [review]

**Review by tahmed-dev — APPROVED** [review]

**Pull request merged**